clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java

```diff
@@ -113,54 +132,99 @@ public PrototypeAsyncConsumer(final Map<String, Object> configs,
         this.defaultApiTimeoutMs = config.getInt(ConsumerConfig.DEFAULT_API_TIMEOUT_MS_CONFIG);
         this.logContext = createLogContext(config, groupRebalanceConfig);
         this.log = logContext.logger(getClass());
-        this.deserializers = new Deserializers<>(config, keyDeserializer, valueDeserializer);
+        Deserializers<K, V> deserializers = new Deserializers<>(config, keyDeserializer, valueDeserializer);
         this.subscriptions = createSubscriptionState(config, logContext);
         this.metrics = createMetrics(config, time);
         List<ConsumerInterceptor<K, V>> interceptorList = getConfiguredConsumerInterceptors(config);
+        this.interceptors = new ConsumerInterceptors<>(interceptorList);
         ClusterResourceListeners clusterResourceListeners = ClientUtils.configureClusterResourceListeners(metrics.reporters(),
                 interceptorList,
                 Arrays.asList(deserializers.keyDeserializer, deserializers.valueDeserializer));
-        this.eventHandler = new DefaultEventHandler(
+        this.metadata = new ConsumerMetadata(config, subscriptions, logContext, clusterResourceListeners);
+        // Bootstrap the metadata with the bootstrap server IP address, which will be used once for the subsequent
+        // metadata refresh once the background thread has started up.
+        final List<InetSocketAddress> addresses = ClientUtils.parseAndValidateAddresses(config);
+        metadata.bootstrap(addresses);
+
+        this.applicationEventQueue = new LinkedBlockingQueue<>();
+        this.backgroundEventQueue = new LinkedBlockingQueue<>();
+
+        this.defaultBackgroundThread = new DefaultBackgroundThread<>(logContext,
+                time,
                 config,
-                groupRebalanceConfig,
-                logContext,
-                subscriptions,
+                applicationEventQueue,
+                backgroundEventQueue,
+                metadata,
                 new ApiVersions(),
-                this.metrics,
-                clusterResourceListeners,
-                null // this is coming from the fetcher, but we don't have one
-        );
-    }
-
-    // Visible for testing
-    PrototypeAsyncConsumer(Time time,
-                           LogContext logContext,
-                           ConsumerConfig config,
-                           SubscriptionState subscriptions,
-                           EventHandler eventHandler,
-                           Metrics metrics,
-                           Optional<String> groupId,
-                           int defaultApiTimeoutMs) {
-        this.time = time;
+                metrics,
+                subscriptions,
+                groupRebalanceConfig);
+        String ioThreadName = NETWORK_THREAD_PREFIX + " | " + getClientId(config);
+        this.ioThread = new KafkaThread(ioThreadName, this.defaultBackgroundThread, true);
+        this.ioThread.start();
```

Comment:
uhm always tricky to start threads in the constructor. Probably sensible to move this to the end of the constructor at least to avoid weird behaviors with initialization order

Reply:
I have moved the thread starting logic elsewhere, but we should discuss.